lib/cext/preprocess.rb

```diff
@@ -131,20 +153,268 @@ def write_array(name, leaking)
+        replacement: 'va_list args; rb_str_cat2(rb_tr_managed_from_handle_or_null(ctx), "Generic error"); return;'
+      }
+    ]
+  },
+  'html_document.c' => {
+    gem: 'nokogiri',
+    patches: [
+      cast_value_for_native('error_list'),
+      cast_value_for_native('io')
+
+    ]
+  },
+  'html_sax_push_parser.c' => {
+    gem: 'nokogiri',
+    patches: [
+      tuple_new_patch('ctx', 'self')
+    ]
+  },
+  'xml_sax_push_parser.c' => {
+    gem: 'nokogiri',
+    patches: [
+      tuple_new_patch('ctx', 'self')
+    ]
+  },
+  'xml_document.c' => {
+    gem: 'nokogiri',
+    patches: [
+      cast_value_for_native('error_list'),
+      cast_value_for_native('io'),
+      cast_native_for_value('ctx'),
+      cast_value_for_native('rb_block_proc()', ';'),
+      {
+        match:'tuple->doc = rb_doc',
+        replacement: 'tuple->doc = rb_tr_handle_for_managed(rb_doc)'
+      },
+      {
+        match: 'tuple->node_cache = cache',
+        replacement: 'tuple->node_cache = rb_tr_handle_for_managed(cache)'
+      },
+      {
+        match: NOKOGIRI_DEALLOC_DECL_ORIG,
+        replacement: NOKOGIRI_DEALLOC_DECL_NEW
+      },
+      {
+        match: NOKOGIRI_DOC_RUBY_OBJECT_ORIG,
+        replacement: NOKOGIRI_DOC_RUBY_OBJECT_NEW
+      },
+      {
+        match: NOKOGIRI_DOC_NODE_CACHE_ORIG,
+        replacement: NOKOGIRI_DOC_NODE_CACHE_NEW
       }
     ]
   },
+  'xml_dtd.c' => {
+    gem: 'nokogiri',
+    patches: [
+      cast_value_for_native('error_list'),
+      cast_native_for_value('data', ';'),
+      cast_value_for_native('hash')
+    ]
+  },
+  'xml_node.c' => {
+    gem: 'nokogiri',
+    patches: [
+      cast_value_for_native('err', ','),
+      cast_value_for_native('error_list'),
+      cast_value_for_native('io'),
+      {
+        match: NOKOGIRI_DOC_RUBY_OBJECT_ORIG,
+        replacement: NOKOGIRI_DOC_RUBY_OBJECT_NEW
+      },
+      {
+        match: NOKOGIRI_DOC_NODE_CACHE_ORIG,
+        replacement: NOKOGIRI_DOC_NODE_CACHE_NEW
+      },
+      cast_native_for_value('node->_private', ';'),
+      cast_value_for_native('rb_node'),
+   ]
+  },
+  'xml_namespace.c' => {
+    gem: 'nokogiri',
+    patches: [
+      { match: NOKOGIRI_DOC_RUBY_OBJECT_ORIG,
+                                  replacement: NOKOGIRI_DOC_RUBY_OBJECT_NEW
+                                },
+      cast_value_for_native('ns'),
+      cast_native_for_value('node->_private', ';')
+    ]
+  },
+  'xml_reader.c' => {
+    gem: 'nokogiri',
+    patches: [
+      cast_value_for_native('error_list'),
+      cast_value_for_native('io'),
+      cast_value_for_native('rb_io')
+    ]
+  },
+  'xml_sax_parser.c' => {
+    gem: 'nokogiri',
+    patches: [
+      {
+        match: 'NOKOGIRI_SAX_SELF(ctx)',
+        replacement: 'rb_tr_managed_from_handle_or_null(NOKOGIRI_SAX_SELF(ctx))'
+      },
+      {
+        match: /va_list args;[^}]*id_warning, 1, ruby_message\);/,
+        replacement: 'rb_funcall(doc, id_warning, 1, NOKOGIRI_STR_NEW2("Warning."));'
```

Comment:
Why is this needed? `vasprintf` doesn't work? Can we fix it?

Reply:
Var args don't work for callbacks from native code to Sulong. I discussed this a bit with Roland and we couldn't see a good way to make it work without some major work to both Sulong and our libffi code.

Reply:
In that case, could we use the native vasprintf and force every argument to go native?